backend/persistent/datastores/csv.js

```diff
@@ -5,81 +5,129 @@ import {type} from 'ramda';
+
+    getAvailableSize: getAvailableSize,
+    setStorageSize: setStorageSize
+};
 
 /**
- *  Error thrown by CSV connector
- *  @class
- *  @param {string}      url    URL of the CSV file that triggered the error
- *  @param {PapaError[]} errors List of errors returned by Papaparse
+ * @typedef {object} CSVStorage Global storage of CSV connector
+ *
+ * @property {Object.<string, object>}
+ *                    data   Store of CSV files parsed into JS objects and indexed by URL
+ * @property {number} size   Available size in bytes (0 to disable limit)
+ * @property {number} used   Used size in bytes
  */
-export function CSVError(url, errors) {
-      /**
-       * Error class
-       * @type {string}
-       */
-      this.name = 'CSVError';
-
-      /**
-       * Error description
-       * @type {string}
-       */
-      this.message = 'Failed to parse CSV file ' + url;
-
-      if (Error.captureStackTrace) {
-          Error.captureStackTrace(this, CSVError);
-      } else {
-          /**
-           * Error stack trace
-           */
-          this.stack = new Error(this.message).stack;
-      }
-
-      /**
-       * URL to CSV file
-       * @type {string}
-       */
-      this.url = url;
-
-      /**
-       * List of errors returned by Papaparse
-       * @type {PapaError[]}
-       */
-      this.errors = errors;
-
-      if (errors && errors[0] && errors[0].message) {
-          this.message = errors[0].message;
-      }
-}
-CSVError.prototype = Object.create(Error.prototype);
-CSVError.prototype.constructor = CSVError;
 
 /**
- * Store of CSV files parsed into JS objects and indexed by URL
+ * CSV-connector global storage.
  *
- * @const {Object.<string, object>}
+ * Note the difference between:
+ * - `const CSV = require('./csv');`
+ * - `import * as CSV from './csv';`
+ * Files using `require` will share STORAGE.
+ * Files using `import` will have their own copy of STORAGE.
+ *
+ * @const {CSVStorage}
  */
-const connectionData = {};
+const STORAGE = {
+    data: {},
+    size: 0,
+    used: 0
+};
+
 function getData(connection) {
-    return connectionData[connection.database];
+    return STORAGE.data[connection.database];
 }
+
 function putData(connection, data) {
-    connectionData[connection.database] = data;
+    const size = connection.size || 0;
+    STORAGE.used += size;
+    STORAGE.data[connection.database] = data;
+}
+
+function deleteData(connection) {
+    const size = connection.size || 0;
+    STORAGE.used = Math.max(0, STORAGE.used - size);
+    delete STORAGE.data[connection.database];
 }
 
-export function connect(connection) {
+function setStorageSize(size) {
+    STORAGE.size = size;
+}
+
+function getAvailableSize() {
+    // no size limit
+    if (STORAGE.size === 0) {
+        return 0;
+    }
+
+    const available = STORAGE.size - STORAGE.used;
+
+    // storage is full
+    if (available <= 0) {
+        return -1;
+    }
+
+    // available size
+    return available;
+}
+
+
+/**
```

Comment:
It looks like you have two comment blocks for this function

Reply:
The first block declares the type `CSVConnection` and the second block declares the function `connect`. See that `CSVConnection` is also used in the declarations of `disconnect`, `query`, `tables` and `schemas`.